packages/storybook/.storybook/preview.js

```diff
@@ -36,6 +36,30 @@ applyPolyfills().then(() => {
   window.CustomEvent = CustomEvent;
 })();
 
+/**
+ * The custom event locationchange will dispatch when the URL changes.
+ * The native events `hashchange` and `popstate` were not reliable.
+ */
+ (() => {
+  const oldPushState = history.pushState;
+  history.pushState = function pushState() {
+    const ret = oldPushState.apply(this, arguments);
+    window.dispatchEvent(new Event('locationchange'));
+    return ret;
+  };
+
+  const oldReplaceState = history.replaceState;
+  history.replaceState = function replaceState() {
+    const ret = oldReplaceState.apply(this, arguments);
+    window.dispatchEvent(new Event('locationchange'));
+    return ret;
+  };
```

Comment:
Can you leave a comment to explain what's going on with the `.apply()`? Also, I don't understand why we are returning anything - what is that for?

Reply:
The approach here is that we are using the native `pushState` and `replaceState` functions so that we can add our own custom event `locationchange` as part of it. So `apply` is making sure the original context and arguments (if any) are still getting used. This solution is based on [the following here](https://stackoverflow.com/questions/6390341/how-to-detect-if-url-has-changed-after-hash-in-javascript#answer-52809105) which I have simplified to only dispatch `locationchange` instead of the other extra events we don't really need. And it's returning the state into the history as it's just kind of a "passthrough" for the history values.

Reply:
It looks like [`.replaceState()` isn't meant to return anything](https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState#return_value)

Reply:
Oh good catch. I didn't notice that in the docs. I think that we can just simplify this even further then to something like this (same for replaceState too):

```js
const pushState = history.pushState;
history.pushState = function () {
  pushState.apply(this, arguments);
  window.dispatchEvent(new Event('locationchange'));
};
```